MetaMask/metamask-mobile
# fix: refactor saga xhr injection and add unit tests

## **Description**
The basic functionality toggle saga relies on monkeypatching the root `XmlHttpRequest.send` and `XmlHttpRequest.open` in order to properly respect user privacy preferences.

But is requires the use of `xhr2` as the unit tests are running in node.js environment with Jest and not in a React Native one.

This PR adds unit tests for the xmlHttpRequestOverride and makes sure the replacement for xhr is properly done according to the environment.

- Break out xhr override functionality from saga into new file  `app/store/sagas/xmlHttpRequestOverride.ts`
- add unit tests for the xmlHttpRequestOverride
- Guard monkeypatching using `xhr2` to only run in Jest environment
- fix: Don't ignore `async`,`username`,`password` options.


## **Related issues**

- #9584 

## **Manual testing steps**

0. Monitor outgoing network requests
1. Toggle basic functionality off
2. Navigate around the app
3. Notice blocked network requests not being sent out
4. Enable basic functionaliy toggle
5. Navigate around the app
6. Notice previously blocked requests being sent out

## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
